### ISSUE TITLE
Feature requests / possible bugs

### ISSUE THREAD

**Issue by akrabu** [machdyne/blaustahl#3]
Hello!  Loving my Blaustahl so far. 

### A few feature suggestions for the built-in text editor...

- CTRL+A: Go to the beginning of the line
- CTRL+E: Go to the end of the line
- CTRL+K:  Delete entire line
- PGUP: Move cursor to top, then if pressed again (or already on bottom), go to the next page up
- PGDN: Move cursor to bottom, then if pressed again (or already on bottom), go to the next page down
- When pressing ⬅️ at the beginning of a line, it would be nice if it would go to the end of the prior line
- When pasting information, it would be nice if it would erase the entire line.  If you have "abcdefg" and paste "123" with the cursor on "a" you'll end up with "123defg".  It would be helpful if it would go ahead and nullify the rest of the line as well.

### Bugs in the text editor

- I'm running v0.0.4 and when I press CTRL+S to hide the status bar, it doesn't appear to do anything?
- When on the last column #80 on any line, if you type a character, you can't backspace it out.  Only the delete key will remove it - except a lot of keyboards don't have a delete key.
- Backspacing out the final line removes one character at the end of the previous line: https://v.usetapes.com/OwkU6QexWE

**Comment by machdyne** [comment]
The CTRL-S issue might be related to #2 which can be resolved with:

```
$ stty -ixon
```

Ultimately it might be necessary to change CTRL-S for toggling the status bar to something else. I will also consider your other key requests. One thing I want to be cautious about is making it too easy to accidentally delete something, but maybe we'll add undo at some point and it won't matter as much.

Thanks for reporting these bugs and feature requests.

**Comment by machdyne** [comment]
Some changes in 0.0.5:

The backspace issue should be fixed.

Either CTRL-S or CTRL-Q (won't freeze screen) will now toggle the status bar.

CTRL-A and CTRL-E should work in read/write and read-only modes.

CTRL-X should now act like the delete key when in write mode.

**Comment by akrabu** [comment]
> One thing I want to be cautious about is making it too easy to accidentally delete something, but maybe we'll add undo at some point and it won't matter as much.

I agree, that would be unfortunate.  I almost requested a way to wipe a whole page and thought better of it.  Deleting just one line seemed like a better alternative - but an undo would make that even better!  OR... make it work like CTRL-K in Nano.  It doesn't _delete_ the line, it _cuts_ it.  So, worst case scenario, you could just paste it back (CTRL-U).  Sooo maybe cut/paste support would be a good alternative?

> Some changes in 0.0.5:

Hmm..  I still can't backspace: https://v.usetapes.com/pRCWukOjZg (apologies for my dishwasher in the background lol)

Neither CTRL-S nor CTRL-Q removed the status bar for me at first, but both work after running `stty -ixon`.

CTRL-A appears to get intercepted by `screen`.  Didn't think of that when I made the request 😅 CTRL-E works, though! 

CTRL-X also works fine 👍 

Thanks!

---

Ps. I tried using `minicom`, and CTRL-A and CTRL-E work.  But now I can't exit Minicom because I'm supposed to hit CTRL-A then X to exit, apparently haha.  CTRL-S/Q also work.  CTRL-X works fine as well.  Buuut PGUP/DN don't work, and I can't change pages.

**Comment by machdyne** [comment]
Because screen uses CTRL-A you need to press CTRL-A then A.

Similarly, CTRL-A S should toggle the status bar.

In Minicom, try CTRL-A CTRL-A.

PGUP/PGDN is working for me in Minicom, and CTRL-A then Q to quit.

I wonder what result you get from pageup/down after running:

```
$ showkey -a
```

With backspace, I only fixed the previous line bug and didn't change the behavior. Backspace only removes the character to the left of the cursor. To delete the last character on a line you'd still need to use delete or CTRL-X.

I think cut/paste would be useful, maybe paste would only be allowed into a blank area.

**Comment by akrabu** [comment]
CTRL-A, A just says "no other window": https://v.usetapes.com/F24FRscS0N

CTRL-A, S doesn't appear to do anything.

CTRL-A, Q doesn't appear to work in Minicom - I have to "killall minicom" in another window.  https://v.usetapes.com/kSokygECNJ

Showkey doesn't appear to capture PGUP/PGDN - it just pages up and down in the terminal as usual:  https://v.usetapes.com/sPwQ8izPqG

Neither PGUP nor PGDN appear to work properly in Minicom.  I think it's just paging up and down in the Terminal window itself:  https://v.usetapes.com/EMLqqKJDqJ

Ah yes, I can't reproduce the backspace bug - thanks!

**Comment by akrabu** [comment]
I got the app [Serial](https://www.decisivetactics.com/products/serial/) and everything seems to work in that.  It even automatically finds my Blaustahl.  When I first load it, the screen is blank, but a PgDn or CTRL+L gets it to redraw the screen.  👍

![Screenshot 2024-06-25 at 6 11 34 PM](https://github.com/machdyne/blaustahl/assets/6868732/9fdb4248-0e83-44c1-b55e-305c254f3201)

Still can't get `minicom` or `screen` to hide the status bar, though.

Ps. I love this little thing.  May have to check out some of your computers!

**Comment by machdyne** [comment]
I just tested screen on OS X and was able to toggle the status bar using: [ctrl-a] [release]  then 's'.

**Comment by akrabu** [comment]
Ohhhh I got it.  I was doing CTRL+A, CTRL+S - but needed to do CTRL+A, S.  Thanks!

**Comment by akrabu** [comment]
To tack one more on, since it is kind of related to these - I noticed if I disable the status bar with ctrl+q, suddenly ctrl+a and ctrl+e no longer appear to jump to the beginning/end of the line.  If I re-enable the status bar with ctrl+q, it will work again.  https://v.usetapes.com/fd483uevzX

And here's a weird one.  If you do ctrl+k, it appears to just move the carat one space down, but in fact it's erasing one space to the right.  See here! https://v.usetapes.com/hqIAGo9rWI

Just found one more.  If you're at the end of the line, type something, and hit backspace - it appears to erase it.  If you refresh the screen, you realize it has not.  This ONLY occurs when the status bar is turned off.  https://v.usetapes.com/RDQlcPuU2T

More statusbar-off shenanigans...  if you paste more characters than can fit at the end of a line, it appears to line wrap... but it has not.  Data loss ensues.  https://v.usetapes.com/6FNhEan7Md

One more - you can't type the [ character, status bar on or off.
